base/compiler/typeinfer.jl

```diff
@@ -470,12 +488,12 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
     end
     if frame === false
         # completely new
-        mi.inInference = true
+        lock_mi_inference(interp, mi)
```

Comment:
I don't know that this is a lock in the proper sense. It's a bootstrapping device, but it's also merely an advisory hint.

Reply:
It's not a lock in the atomic sense, but I do think it's a semantic lock.